crates/functional/src/lib.rs

```diff
@@ -181,6 +182,194 @@ where
     );
 }
 
+pub fn use_effect<F, Destructor>(callback: F)
+where
+    F: FnOnce() -> Destructor + 'static,
+    Destructor: FnOnce() + 'static,
+{
+    let callback = Box::new(callback);
+    struct UseEffectState<Destructor> {
+        destructor: Option<Box<Destructor>>,
+    }
+    use_hook(
+        |_: &mut UseEffectState<Destructor>, hook_update| {
+            return move || {
+                hook_update(move |state: &mut UseEffectState<Destructor>| {
+                    if let Some(de) = state.destructor.take() {
```

Comment:
We need to make sure that the destructor is called when the component is destroyed

Reply:
Oh right yes.